Add prometheus_exporter seperate container

kmahern opened the pull request:
#### What
Enable prometheus_exporter gem on VCD

#### Ticket
https://dsdmoj.atlassian.net/browse/CBO-1586

#### Why
Currently we are using the prometheus-client gem (https://github.com/prometheus/client_ruby). I believe this does not offer as rich a set of default metrics as prometheus_exporter, or as much flexibility over implementation.

#### How
Added prometheus_exporter gem to Gemfile and configured in prometheus_metrics.rb

Specified a new container in deployment.yaml. Run prometheus_exporter on the default port of 9394 in this container (bundle exec prometheus_exporter --bind 0.0.0.0). Amended service.yaml to expose this port with the name 'metrics'. This is the port that will be scraped by Cloud Platform's Prometheus instance (defined in service-monitor.yaml in the cloud platform environments repo).  Amended deploy scripts to apply the relevant docker image to the new container.

kmahern commented:
@jsugarman in this pr I am trying to run prometheus exporter in a separate container and on a separate port (9394), but within the same pod as the app. The deployment fails and if I query the (container) logs I see the error `container "laa-court-data-ui-metrics" in pod "laa-court-data-ui-b449fbf97-9mvw6" is waiting to start: image can't be pulled` - was wondering if you had any ideas what I am doing wrong?

kmahern commented:
I have updated deploy.sh to set the images for both containers independently kubectl `set image -f .k8s/${environment}/deployment.yaml laa-court-data-ui-app=${docker_image_tag} laa-court-data-ui-metrics=${docker_image_tag} --local -o yaml | kubectl apply -f -` this appears to work, however, now the preexisting pods are not deleted as part of the circleci run - meaning that the new pods cannot be created correctly?

kmahern commented:
All the pods/containers now appear to start correctly and liveness/readiness probes work.

However when I run curl https://dev.view-court-data.service.justice.gov.uk:9394/metrics nothing is returned and I see in the application logs:-

```
I, [2020-11-11T09:13:35.630380 #17]  INFO -- : [57e6fd26-2081-4c9d-ba8c-a6ffc706dafb] Started GET "/metrics" for 100.100.97.201 at 2020-11-11 09:13:35 +0000
I, [2020-11-11T09:13:35.631063 #17]  INFO -- : [57e6fd26-2081-4c9d-ba8c-a6ffc706dafb] Processing by ErrorsController#not_found as HTML
I, [2020-11-11T09:13:35.631112 #17]  INFO -- : [57e6fd26-2081-4c9d-ba8c-a6ffc706dafb]   Parameters: {"path"=>"metrics"}
I, [2020-11-11T09:13:35.631597 #17]  INFO -- : [57e6fd26-2081-4c9d-ba8c-a6ffc706dafb]   Rendering errors/not_found.html.haml within layouts/application
I, [2020-11-11T09:13:35.632054 #17]  INFO -- : [57e6fd26-2081-4c9d-ba8c-a6ffc706dafb]   Rendered errors/not_found.html.haml within layouts/application (Duration: 0.4ms | Allocations: 147)
I, [2020-11-11T09:13:35.632891 #17]  INFO -- : [57e6fd26-2081-4c9d-ba8c-a6ffc706dafb]   Rendered layouts/_google_analytics.html.haml (Duration: 0.1ms | Allocations: 44)
I, [2020-11-11T09:13:35.633338 #17]  INFO -- : [57e6fd26-2081-4c9d-ba8c-a6ffc706dafb]   Rendered layouts/_cookie_banner.html.haml (Duration: 0.3ms | Allocations: 121)
I, [2020-11-11T09:13:35.633556 #17]  INFO -- : [57e6fd26-2081-4c9d-ba8c-a6ffc706dafb]   Rendered layouts/_govuk_logo.html.haml (Duration: 0.1ms | Allocations: 45)
I, [2020-11-11T09:13:35.634120 #17]  INFO -- : [57e6fd26-2081-4c9d-ba8c-a6ffc706dafb]   Rendered layouts/_phase_banner.html.haml (Duration: 0.2ms | Allocations: 90)
I, [2020-11-11T09:13:35.634664 #17]  INFO -- : [57e6fd26-2081-4c9d-ba8c-a6ffc706dafb]   Rendered layouts/_footer.html.haml (Duration: 0.2ms | Allocations: 204)
I, [2020-11-11T09:13:35.634908 #17]  INFO -- : [57e6fd26-2081-4c9d-ba8c-a6ffc706dafb] Completed 404 Not Found in 4ms (Views: 3.5ms | Allocations: 1758)
I, [2020-11-11T09:13:41.364952 #17]  INFO -- : [92d920f4-7664-488b-8abe-3c34e5e3bc4c] Started GET "/ping.json" for 172.20.101.106 at 2020-11-11 09:13:41 +0000
```

jsugarman reviewed (commented):
if we are going with a separate container in the same pod it would seem to make more sense for this container to be in the worker pod rather than the app pod. Have you tried that?

jsugarman commented on .k8s/dev/deployment.yaml:
```diff
@@ -110,3 +110,24 @@ spec:
+              port: 9394
+            initialDelaySeconds: 10
+            periodSeconds: 60
+          env:
```
probably not be needed either

jsugarman commented on .k8s/production/deployment.yaml:
```diff
@@ -113,3 +113,24 @@ spec:
+              port: 9394
+            initialDelaySeconds: 10
+            periodSeconds: 60
+          env:
```
[ditto](https://github.com/ministryofjustice/laa-court-data-ui/pull/332/files#r522060924)

jsugarman commented on .k8s/staging/deployment.yaml:
```diff
@@ -113,3 +113,24 @@ spec:
+              port: 9394
+            initialDelaySeconds: 10
+            periodSeconds: 60
+          env:
```
[ditto](https://github.com/ministryofjustice/laa-court-data-ui/pull/332/files#r522060924)

jsugarman commented on .k8s/uat/deployment.yaml:
```diff
@@ -115,3 +115,24 @@ spec:
+              port: 9394
+            initialDelaySeconds: 10
+            periodSeconds: 60
+          env:
```
[ditto](https://github.com/ministryofjustice/laa-court-data-ui/pull/332/files#r522060924)

jsugarman commented on config/initializers/prometheus_metrics.rb:
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+if Rails.env.production?
```
**not a blocker**  this could run locally if rails server startup included a `bundle exec prometheus_exporter` somehow - could use MakeFile or `config.ru` for example

jsugarman replied:
naming wise  I'd call this file just `prometheus.rb`

jsugarman reviewed (changes requested):
Can you add a readme entry in some logical place and i'd rename the initializer to just `prometheus`. But other than that and barring the our discussions around separate pods or worker pods containers i'd say its good to go.

kmahern commented:
> if we are going with a separate container in the same pod it would seem to make more sense for this container to be in the worker pod rather than the app pod. Have you tried that?

I have tried this here https://github.com/ministryofjustice/laa-court-data-ui/pull/348, however, it didn't work. In the application logs I see the error `Prometheus Exporter, failed to send message Address not available - connect(2) for "localhost" port 9394`

kmahern commented:
> Can you add a readme entry in some logical place and i'd rename the initializer to just `prometheus`. But other than that and barring the our discussions around separate pods or worker pods containers i'd say its good to go.

👍 I have renamed the initializer and added a readme entry (let me know if you think there is anything missing).

jsugarman commented on docs/build_and_deploy.md:
```diff
@@ -0,0 +1,15 @@
+
+### Prometheus
+
+The prometheus_exporter gem is used gather application metrics.
```
```suggestion
The prometheus_exporter gem is used to gather application metrics.
```

kmahern replied:
👍 Have made suggested change.

jsugarman commented on docs/build_and_deploy.md:
```diff
@@ -0,0 +1,15 @@
+## Build and Deploy
```
This is great but build and deploy should probably be left for orchestration and config related to building and deploying (like [CCCDS build and deploy doc](https://github.com/ministryofjustice/Claim-for-Crown-Court-Defence/blob/master/docs/build_and_deploy.md)).

This would be betters as its on doc. I would suggest just renaming this `docs/monitoring.md` (and amend title)

kmahern replied:
👍 Have renamed to docs/monitoring.md and amended title

jsugarman reviewed (approved):
👍 great
